### ISSUE TITLE
btrbk gobbles existing subvolumes on "localhost" targets

### ISSUE THREAD

**Issue by Witiko** [digint/btrbk#233]
At the beginning, the `/media/backups/rootfs.20180506` subvolume exists on the remote host. After running `btrbk resume`, Btrbk attempts to perform an incremental update, fails with the “File exists” error and deletes the existing subvolume:
```
$ git rev-parse HEAD
c8fcb01bc23f8ebc2cb54d015441b954e971cf88
# ./btrbk resume
WARNING: [send/receive] (send=/media/witiko/rootfs/btrbk_snapshots/rootfs.20180506, receive=/media/backups) At subvol /media/witiko/rootfs/btrbk_snapshots/rootfs.20180506
WARNING: [send/receive] (send=/media/witiko/rootfs/btrbk_snapshots/rootfs.20180506, receive=/media/backups) At subvol rootfs.20180506
ERROR: creating subvolume rootfs.20180506 failed: File exists
ERROR: Failed to send/receive btrfs subvolume: /media/witiko/rootfs/btrbk_snapshots/rootfs.20180506  -> localhost:/media/backups
WARNING: Deleted partially received (garbled) subvolume: localhost:/media/backups/rootfs.20180506
```
In the following invocation of `btrbk resume`, the entire `rootfs.20180506` subvolume is transferred non-incrementally, wasting storage space and network bandwidth.

**Comment by digint** [comment]
I assume that `rootfs.20180506` was incompletely transferred (garbled) by an earlier btrbk send/receive (if it was not, btrbk would not have tried to send/receive it). In this case, deleting the garbled target subvolume is the correct behavior. Do you have any logs of earlier btrbk calls (e.g. transaction-log) ?

Now what embarrasses me more is why it came to a situation where no common parents can be found, triggering a full non-incremental send/receive. What is your strategy here? How do you create / delete your snapshots? What was the last command that deleted snapshots and/or target subvolumes?

**Comment by Witiko** [comment]
> I assume that `rootfs.20180506` was incompletely transferred (garbled) by an earlier btrbk send/receive (if it was not, btrbk would not have tried to send/receive it). In this case, deleting the garbled target subvolume is the correct behavior. Do you have any logs of earlier btrbk calls (e.g. transaction-log)?

Today at 2AM, btrbk created new snapshots and backed them up. However, btrbk also decided to send snapshots from 2018/04/02, although they already existed at the target. The sending failed (likely with an `File exists` error again, the logs do not say) and the target volumes were deleted by btrbk.

```sh
$ sudo grep btrbk /var/log/syslog
May 29 02:00:01 inspiron CRON[31515]: (root) CMD (set -e; chronic /opt/btrbk/btrbk clean; chronic /opt/btrbk/btrbk run)
May 29 02:00:04 inspiron btrbk: startup v0.27.0-dev - - - # btrbk command line client, version 0.27.0-dev
May 29 02:00:04 inspiron btrbk: finished success - - - -
May 29 02:00:07 inspiron btrbk: startup v0.27.0-dev - - - # btrbk command line client, version 0.27.0-dev
May 29 02:00:07 inspiron btrbk: snapshot starting /media/witiko/rootfs/btrbk_snapshots/rootfs.20180529 /media/witiko/rootfs/rootfs - -
May 29 02:00:07 inspiron btrbk: snapshot success /media/witiko/rootfs/btrbk_snapshots/rootfs.20180529 /media/witiko/rootfs/rootfs - -
May 29 02:00:07 inspiron btrbk: snapshot starting /media/witiko/home/btrbk_snapshots/home.20180529 /media/witiko/home/home - -
May 29 02:00:09 inspiron btrbk: snapshot success /media/witiko/home/btrbk_snapshots/home.20180529 /media/witiko/home/home - -
May 29 02:00:09 inspiron btrbk: snapshot starting /media/witiko/home/btrbk_snapshots/photography.20180529 /media/witiko/home/home/witiko/images/photography - -
May 29 02:00:10 inspiron btrbk: snapshot success /media/witiko/home/btrbk_snapshots/photography.20180529 /media/witiko/home/home/witiko/images/photography - -
May 29 02:00:10 inspiron btrbk: snapshot starting /media/witiko/home/btrbk_snapshots/documents.20180529 /media/witiko/home/home/witiko/documents - -
May 29 02:00:11 inspiron btrbk: snapshot success /media/witiko/home/btrbk_snapshots/documents.20180529 /media/witiko/home/home/witiko/documents - -
May 29 02:00:11 inspiron btrbk: send-receive starting ssh://localhost/media/backups/rootfs.20180402 /media/witiko/rootfs/btrbk_snapshots/rootfs.20180402 - -
May 29 02:00:23 inspiron btrbk: send-receive ERROR ssh://localhost/media/backups/rootfs.20180402 /media/witiko/rootfs/btrbk_snapshots/rootfs.20180402 - -
May 29 02:00:23 inspiron btrbk: delete_garbled starting ssh://localhost/media/backups/rootfs.20180402 - - -
May 29 02:00:36 inspiron btrbk: delete_garbled success ssh://localhost/media/backups/rootfs.20180402 - - -
May 29 02:00:36 inspiron btrbk: abort_target ABORT ssh://localhost/media/backups - - # Failed to send/receive subvolume
May 29 02:00:36 inspiron btrbk: send-receive starting ssh://localhost/media/backups/home.20180402 /media/witiko/home/btrbk_snapshots/home.20180402 - -
May 29 02:00:39 inspiron btrbk: send-receive ERROR ssh://localhost/media/backups/home.20180402 /media/witiko/home/btrbk_snapshots/home.20180402 - -
May 29 02:00:39 inspiron btrbk: delete_garbled starting ssh://localhost/media/backups/home.20180402 - - -
May 29 02:00:41 inspiron btrbk: delete_garbled success ssh://localhost/media/backups/home.20180402 - - -
May 29 02:00:41 inspiron btrbk: abort_target ABORT ssh://localhost/media/backups - - # Failed to send/receive subvolume
May 29 02:00:41 inspiron btrbk: send-receive starting ssh://localhost/media/backups/photography.20180305 /media/witiko/home/btrbk_snapshots/photography.20180305 - -
May 29 02:00:46 inspiron btrbk: send-receive ERROR ssh://localhost/media/backups/photography.20180305 /media/witiko/home/btrbk_snapshots/photography.20180305 - -
May 29 02:00:46 inspiron btrbk: delete_garbled starting ssh://localhost/media/backups/photography.20180305 - - -
May 29 02:00:48 inspiron btrbk: delete_garbled success ssh://localhost/media/backups/photography.20180305 - - -
May 29 02:00:48 inspiron btrbk: abort_target ABORT ssh://localhost/media/backups - - # Failed to send/receive subvolume
May 29 02:00:48 inspiron btrbk: send-receive starting ssh://localhost/media/backups/documents.20180205 /media/witiko/home/btrbk_snapshots/documents.20180205 - -
May 29 02:00:50 inspiron btrbk: send-receive ERROR ssh://localhost/media/backups/documents.20180205 /media/witiko/home/btrbk_snapshots/documents.20180205 - -
May 29 02:00:50 inspiron btrbk: delete_garbled starting ssh://localhost/media/backups/documents.20180205 - - -
May 29 02:00:53 inspiron btrbk: delete_garbled success ssh://localhost/media/backups/documents.20180205 - - -
May 29 02:00:53 inspiron btrbk: abort_target ABORT ssh://localhost/media/backups - - # Failed to send/receive subvolume
May 29 02:00:53 inspiron btrbk: finished partial - - - # At least one backup task aborted
```

Note that btrbk did not delete or try to send the snapshots from 2018/04/02 during the previous run and there was no manual intervention in the meantime:

```sh
$ sudo grep btrbk /var/log/syslog.1
May 28 02:00:01 inspiron CRON[13296]: (root) CMD (set -e; chronic /opt/btrbk/btrbk clean; chronic /opt/btrbk/btrbk run)
May 28 02:00:12 inspiron btrbk: startup v0.27.0-dev - - - # btrbk command line client, version 0.27.0-dev
May 28 02:00:12 inspiron btrbk: finished success - - - -
May 28 02:00:23 inspiron btrbk: startup v0.27.0-dev - - - # btrbk command line client, version 0.27.0-dev
May 28 02:00:23 inspiron btrbk: snapshot starting /media/witiko/rootfs/btrbk_snapshots/rootfs.20180528 /media/witiko/rootfs/rootfs - -
May 28 02:00:23 inspiron btrbk: snapshot success /media/witiko/rootfs/btrbk_snapshots/rootfs.20180528 /media/witiko/rootfs/rootfs - -
May 28 02:00:23 inspiron btrbk: snapshot starting /media/witiko/home/btrbk_snapshots/home.20180528 /media/witiko/home/home - -
May 28 02:00:25 inspiron btrbk: snapshot success /media/witiko/home/btrbk_snapshots/home.20180528 /media/witiko/home/home - -
May 28 02:00:25 inspiron btrbk: snapshot starting /media/witiko/home/btrbk_snapshots/photography.20180528 /media/witiko/home/home/witiko/images/photography - -
May 28 02:00:26 inspiron btrbk: snapshot success /media/witiko/home/btrbk_snapshots/photography.20180528 /media/witiko/home/home/witiko/images/photography - -
May 28 02:00:26 inspiron btrbk: snapshot starting /media/witiko/home/btrbk_snapshots/documents.20180528 /media/witiko/home/home/witiko/documents - -
May 28 02:00:27 inspiron btrbk: snapshot success /media/witiko/home/btrbk_snapshots/documents.20180528 /media/witiko/home/home/witiko/documents - -
May 28 02:00:27 inspiron btrbk: send-receive starting ssh://10.0.0.11/media/backups/rootfs.20180528 /media/witiko/rootfs/btrbk_snapshots/rootfs.20180528 /media/witiko/rootfs/btrbk_snapshots/rootfs.20180527 -
May 28 02:03:16 inspiron btrbk: send-receive success ssh://10.0.0.11/media/backups/rootfs.20180528 /media/witiko/rootfs/btrbk_snapshots/rootfs.20180528 /media/witiko/rootfs/btrbk_snapshots/rootfs.20180527 -
May 28 02:03:16 inspiron btrbk: send-receive starting ssh://10.0.0.11/media/backups/home.20180528 /media/witiko/home/btrbk_snapshots/home.20180528 /media/witiko/home/btrbk_snapshots/home.20180527 -
May 28 02:04:33 inspiron btrbk: send-receive success ssh://10.0.0.11/media/backups/home.20180528 /media/witiko/home/btrbk_snapshots/home.20180528 /media/witiko/home/btrbk_snapshots/home.20180527 -
May 28 02:04:33 inspiron btrbk: send-receive starting ssh://10.0.0.11/media/backups/photography.20180528 /media/witiko/home/btrbk_snapshots/photography.20180528 /media/witiko/home/btrbk_snapshots/photography.20180527 -
May 28 02:04:42 inspiron btrbk: send-receive success ssh://10.0.0.11/media/backups/photography.20180528 /media/witiko/home/btrbk_snapshots/photography.20180528 /media/witiko/home/btrbk_snapshots/photography.20180527 -
May 28 02:04:42 inspiron btrbk: send-receive starting ssh://10.0.0.11/media/backups/documents.20180528 /media/witiko/home/btrbk_snapshots/documents.20180528 /media/witiko/home/btrbk_snapshots/documents.20180527 -
May 28 02:13:29 inspiron btrbk: send-receive success ssh://10.0.0.11/media/backups/documents.20180528 /media/witiko/home/btrbk_snapshots/documents.20180528 /media/witiko/home/btrbk_snapshots/documents.20180527 -
May 28 02:13:29 inspiron btrbk: delete_snapshot starting /media/witiko/home/btrbk_snapshots/documents.20180512 - - -
May 28 02:13:29 inspiron btrbk: delete_snapshot success /media/witiko/home/btrbk_snapshots/documents.20180512 - - -
May 28 02:13:29 inspiron btrbk: finished success - - - -
```

> What is your strategy here? How do you create / delete your snapshots? What was the last command that deleted snapshots and/or target subvolumes?

I execute `btrbk run` every day at 2AM. The configuration is as follows:

```
$ cat /etc/btrbk/btrbk.conf
snapshot_preserve_min   2d
snapshot_preserve       7d  2w 2m

target_preserve_min     no
target_preserve         20d 10w 3m 1y

snapshot_create         always
snapshot_dir            btrbk_snapshots

transaction_syslog      daemon

ssh_user                witiko
ssh_port                10222
stream_compress         gzip
stream_compress_level   9
rate_limit              100k

volume /media/witiko/rootfs
  subvolume  rootfs
    target send-receive localhost:/media/backups/

volume /media/witiko/home
  subvolume  home
    target send-receive localhost:/media/backups/
  subvolume  home/witiko/images/photography
    target send-receive localhost:/media/backups/
  subvolume  home/witiko/documents
    snapshot_preserve   15d 4w 6m
    target send-receive localhost:/media/backups/
```

**Comment by digint** [comment]
Looks like you are successfully confusing btrbk by using "localhost" in your config. Btrbk assumes that "localhost" has the same mountpoints as your local root, which is probably not the case (I assume you run some virtual machine on localhost:10222).

I pushed a hotfix (847470a5ef9bbfc3d12fd427d600eaa0d1444d02) for this on the [fix-localhost branch](/digint/btrbk/tree/fix-localhost), hope that helps.

**Comment by Witiko** [comment]
That explains why the bug did not manifest itself in 2018/04/02, because `/media/backups` was also mounted locally at that time.

**Comment by Witiko** [comment]
Clearly, caching should take ports into account, but what is the argument for special-casing `localhost`? Forwarding the local TCP port 22 to a remote host is clearly asking for trouble, but it is easy to do. Wouldn't it be less headache to treat `localhost` as any other hostname?

**Comment by digint** [comment]
> Wouldn't it be less headache to treat localhost as any other hostname?

That's exactly what the hotfix does. The point is that I need some (globally unique) key for the mountpoint- and btrfs-tree-cache, and I took `localhost` if there is no host defined in the config (the host running btrbk). After the hotfix this is now `#localhost`.

But the key should really be a `host:port` tuple, e.g. for people running multiple VM's on the same machine with access via different ports. Especially when the result is so disastrous (as it was in your case).

**Comment by digint** [comment]
Added documentation about different ports in 2d0a705e43414e0a85023f0d1f630f105affa63f, as well as sanity checks before autmatical deletion of garbled subvolumes in 9d00a2ac83c96c46077d4fd5afca52eae853edab.

With this fix, btrbk would not have deleted your already existing (read-only) subvolumes on the target host.

On second thought, keeping distinct btrfs-trees for different `host:port` would be more confusing (for the general use case) as useful.

**Comment by Witiko** [comment]
Looks good to me. I am experiencing no issues with 9d00a2a.

**Comment by ekrohn** [comment]
Another note since the hotfix is not in mainline. I ran into a similar problem with ssh://localhost/ and non-standard port. My scenario was not a virtual machine, but an ssh tunnel from my laptop to my desktop at home, where the laptop's port 5022 redirects to the desktop's port 22, and the desktop's port 22 opens only in response to successful fwknop.

In my case, btrbk would never get the details of the remote pool, so would always try to do a full backup to the remote, which never worked after the first time because the target already existed. After spending hours trying to figure out what was broken (it used to work sometime before 0.27.1 or 0.27.0), I found this issue thread.

The hotfix does have a note in btrbk.conf.5.asciidoc for a work around. I mention it here in case someone does a web search and does not dig into the git diff or does not understand the reference to 'alias'.
In /root/.ssh/config, add:

    Host btrbk-remote
        Hostname localhost
        Port <whatever>

then change /etc/btrbk/btrbk.conf to use ssh://btrbk-remote/... instead of ssh://localhost/... and all is well.

**Comment by digint** [comment]
Good news! I found some time to work on this and finally merged it to master:

- d8b7988ffab046611d1b160909eae5a5017baf28 btrbk: deprecate ssh_port option in favor of ssh://hostname[:port] notation
- 2f09a9a72319f9ba7a72fff27970954909391abf btrbk: virtual machine setups; use MACHINE_ID for caches; use port in URL
- ca3bd505effb46cf099fe6299829aac2550114ad btrbk: make check for duplicate snapshot/backup locations aware of shared btr_tree

Notable changes are the ssh port configuration, which can be set directly in `volume` and `target` section, using the `ssh://hostname[:port]` notation, e.g.:

```
volume ssh://source.host.com:2222/mnt/btr_pool
  target ssh://target.bahost.com:2233/mnt/btr_backup
```
See also: https://github.com/digint/btrbk#example-virtual-machine-setup

The `ssh_port` configuration option has been dropped in favour of this, which does not allow ambiguous settings any more.

Btrbk should now handle all cases correctly, even if distinct ports map to the same filesystem, shown by info messages like this:
```
Assuming same filesystem: "ssh://localhost/dev/loop0", "ssh://localhost:22/dev/loop0"
```

@ekrohn @Witiko could you please give it a try, and report back if this all works for your setups?

**Comment by ekrohn** [comment]
It works in my scenario (ssh://localhost:5022/...).
Thank you!

**Comment by Witiko** [comment]
This is excellent news, @digint. I have just pulled the current master and I will let you know if there are any regressions by tomorrow.

**Comment by Witiko** [comment]
I appologize about the delay. I encountered no regressions.

**Comment by digint** [comment]
merged in btrbk-0.28.0, closing issue
